src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +156,33 @@ protected function doDelete(array $ids, array $tagData = []): bool
      */
     protected function doInvalidate(array $tagIds): bool
     {
-        if (!$this->redisServerSupportSPOP()) {
-            return false;
-        }
-
-        // Pop all tag info at once to avoid race conditions
-        $tagIdSets = $this->pipeline(static function () use ($tagIds) {
-            foreach ($tagIds as $tagId) {
-                // Client: Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
-                // Server: Redis 3.2 or higher (https://redis.io/commands/spop)
-                yield 'sPop' => [$tagId, self::POP_MAX_LIMIT];
+        // Rename all sets so we can safely avoid race conditions
+        $this->pipeline(static function () use ($tagIds) {
+            $uniqueToken = bin2hex(random_bytes(10));
+            foreach ($tagIds as $key => $tagId) {
+                $tagIds[$key] = $temporaryTagId = $tagId . ':' . $uniqueToken;
+                yield 'rename' => [$tagId, $temporaryTagId];
             }
         });
 
-        // Flatten generator result from pipeline, ignore keys (tag ids)
-        $ids = array_unique(array_merge(...iterator_to_array($tagIdSets, false)));
+
+        // Now safely take the time to scan the keys in each set and collect ids we need to delete
+        $idHash = [];
+        foreach ($tagIds as $temporaryTagId) {
+            $cursor = null;
+            while(($ids = $this->redis->sScan($temporaryTagId, $cursor, null, self::SSCAN_LIMIT)) !== false) {
```

Comment:
false !== and get rid of the extra brackets

Reply:
Have unpushed CS fix for this, will push once I fix the test failures.

Reply:
> A hypothetical future step could be to schedule the cleanups using Messenger :)

well, we kind of need to at least expire the affected id's immediately. But yes, removing the Set's could be async, but that will normally only be a few keys which is here bulk deleted with the affected ids.